src/pudl/etl/epacems_assets.py

```diff
@@ -148,8 +149,438 @@ def hourly_emissions_epacems(
+    #.  The second issue is that there are many NA subplant_id that we should fill. To
+        do this, we first look at unit_id_pudl. If a group of generators are assigned a
+        unit_id_pudl but have NA subplant_ids, we assign a single new subplant_id to this
+        group of generators. If there are still generators at a plant that have both NA
+        subplant_id and NA unit_id_pudl, we for now assume that each of these generators
+        consitutes its own subplant. We thus assign a unique subplant_id to each
+        generator that is unique from any existing subplant_id already at the plant. In
+        the case that there are multiple emissions_unit_id_epa at a plant that are not
+        matched to any other identifiers (generator_id, unit_id_pudl, or subplant_id), as
+        is the case when there are units that report to CEMS but which do not exist in
+        the EIA data, we assign these units to a single subplant.
+
+    Args:
+        subplant_crosswalk: a dataframe containing the output of
+            :func:`pudl.analysis.epacamd_eia.make_subplant_ids`
+    """
+    # Step 1: Create corrected versions of subplant_id and unit_id_pudl
+    # if multiple unit_id_pudl are connected by a single subplant_id, unit_id_pudl_connected groups these unit_id_pudl together
+    subplant_crosswalk = connect_ids(
+        subplant_crosswalk, id_to_update="unit_id_pudl", connecting_id="subplant_id"
+    )
+    # if multiple subplant_id are connected by a single unit_id_pudl, group these subplant_id together
+    subplant_crosswalk = connect_ids(
+        subplant_crosswalk, id_to_update="subplant_id", connecting_id="unit_id_pudl"
+    )
+
+    # Step 2: Fill missing subplant_id
+    # We will use unit_id_pudl to fill missing subplant ids, so first we need to fill any missing unit_id_pudl
+    # We do this by assigning a new unit_id_pudl to each generator that isn't already grouped into a unit
+
+    # create a numeric version of each generator_id
+    # ngroup() creates a unique number for each element in the group
+    subplant_crosswalk["numeric_generator_id"] = subplant_crosswalk.groupby(
+        ["plant_id_eia", "generator_id"], dropna=False
+    ).ngroup()
+    # when filling in missing unit_id_pudl, we don't want these numeric_generator_id to overlap existing unit_id
+    # to ensure this, we will add 1000 to each of these numeric generator ids to ensure they are unique
+    # 1000 was chosen as an arbitrarily high number, since the largest unit_id_pudl is ~ 10.
+    subplant_crosswalk["numeric_generator_id"] = (
+        subplant_crosswalk["numeric_generator_id"] + 1000
+    )
+    # fill any missing unit_id_pudl with a number for each unique generator
+    subplant_crosswalk["unit_id_pudl_filled"] = (
+        subplant_crosswalk["unit_id_pudl_connected"]
+        .fillna(subplant_crosswalk["subplant_id_connected"] + 100)
+        .fillna(subplant_crosswalk["numeric_generator_id"])
+    )
```

Comment:
I don't fully understand why `numeric_generator_id` could overlap with existing `unit_id` but I feel like there might be a more robust way to ensure they don't overlap. Does this function have access to the `unit_id` values? Could we create `numeric_generator_id`s that don't exist in `unit_id` without just adding an arbitrary number?

Reply:
yeah this stuff was a direct copy/paste from greg that I didn't augment much or spend a lot of time fully understanding it tbqh. But I think we can change the +1000 to be `max(subplant_crosswalk["numeric_generator_id"]) + 1`. i'll double check this...

Reply:
Actually I think this is the way

```
numeric_generator_id_gb = subplant_crosswalk.groupby(
        ["plant_id_eia", "generator_id"], dropna=False
    )
    subplant_crosswalk["numeric_generator_id"] = (
        numeric_generator_id_gb.ngroup() + numeric_generator_id_gb["unit_id_pudl"].max()
    )
```

Reply:
these columns are all interim values that are used to update the subplant_id but not actually published. I'll chew on helpful defensive checks to add in here though.

Reply:
okay i went back in there to add some logs and defensive checks in this function. I learned that there are actually no null `subplant_id` after the current version of `make_subplant_ids`. I believe this is because I condensed the step where IDs were being merged into the crosswalk to be before `make_subplant_ids` instead of sprinkled throughout the process. I did a quick check to make sure that I wasn't missing any records this way and it seems solid:

<img width="1040" alt="image" src="https://user-images.githubusercontent.com/25487998/233698702-3c58092d-deed-4857-971f-5ea1e0aeccf1.png">

soooo I was able to delete some of the fillna steps in this function. Right now it is only doing step one and four of the original documented steps:
```
    High-level overview of method:
        1. Use the PUDL subplant_id if available. In the case where a unit_id_pudl groups several subplants,
        we overwrite these multiple existing subplant_id with a single subplant_id.
        2. Where there is no PUDL subplant_id, we use the unit_id_pudl to assign a unique subplant_id
        3. Where there is neither a pudl subplant_id nor unit_id_pudl, we use the generator ID to
        assign a unique subplant_id
        4. All of the new unique ids are renumbered in consecutive ascending order
```

Reply:
okay i went back in there to add some logs and defensive checks in this function. I learned that there are actually no null `subplant_id` after the current version of `make_subplant_ids`. I believe this is because I condensed the step where IDs were being merged into the crosswalk to be before `make_subplant_ids` instead of sprinkled throughout the process. I did a quick check to make sure that I wasn't missing any records this way and it seems solid:

<img width="1040" alt="image" src="https://user-images.githubusercontent.com/25487998/233698702-3c58092d-deed-4857-971f-5ea1e0aeccf1.png">

soooo I was able to delete some of the fillna steps in this function. Right now it is only doing step one and four of the original documented steps:
```
    High-level overview of method:
        1. Use the PUDL subplant_id if available. In the case where a unit_id_pudl groups several subplants,
        we overwrite these multiple existing subplant_id with a single subplant_id.
        2. Where there is no PUDL subplant_id, we use the unit_id_pudl to assign a unique subplant_id
        3. Where there is neither a pudl subplant_id nor unit_id_pudl, we use the generator ID to
        assign a unique subplant_id
        4. All of the new unique ids are renumbered in consecutive ascending order
```